File: pypp_cli/test_dir/cpp/src/lists/lists.cpp
#include "lists/lists.h"
#include "py_list.h"
#include "py_str.h"
#include "pypp_util/print.h"
#include "pypp_util/to_py_str.h"
#include "slice/creators.h"
#include <utility>

namespace me {
void _test_fn(pypp::PyList<pypp::PyStr> arg1) { pypp::print(arg1); }

void list_fn() {
    pypp::print(pypp::PyStr("LIST RESULTS:"));
    pypp::PyList<int> a({1, 2, 3, 4, 5});
    pypp::print(a);
    a[0] = 10;
    pypp::print(a);
    a.append(11);
    pypp::print(a);
    pypp::print(pypp::to_pystr(a[-1]));
    int b = a[-1];
    pypp::print(pypp::to_pystr(b));
    b = 20;
    pypp::print(pypp::to_pystr(b));
    pypp::print(a);
    pypp::print(pypp::to_pystr(a.len()));
    a.reverse();
    pypp::print(a);
    pypp::print(pypp::to_pystr(a.count(2)));
    pypp::print(pypp::to_pystr(a.index(2)));
    a.remove(2);
    pypp::print(a);
    a.insert(4, 2);
    pypp::print(a);
    int ins_val = 2;
    a.insert(4, std::move(ins_val));
    pypp::print(a);
    int popped_val = a.pop();
    pypp::print(pypp::to_pystr(popped_val));
    pypp::print(a);
    int popped_val2 = a.pop(1);
    pypp::print(pypp::to_pystr(popped_val2));
    pypp::print(a);
    a.clear();
    pypp::print(a);
    pypp::print(pypp::to_pystr(pypp::PyList({1, 2}) == pypp::PyList({1, 2})));
    pypp::print(pypp::to_pystr(pypp::PyList({1, 2}) < pypp::PyList({1, 2})));
    pypp::print(pypp::to_pystr(pypp::PyList({1, 2}) <= pypp::PyList({1, 2})));
    pypp::print(pypp::to_pystr(pypp::PyList({1, 2}) > pypp::PyList({1, 2})));
    pypp::print(pypp::to_pystr(pypp::PyList({1, 2}) >= pypp::PyList({1, 2})));
    pypp::print(pypp::to_pystr(pypp::PyList({1, 2}) != pypp::PyList({1, 2})));
    pypp::PyList<int> c = pypp::PyList({1, 2}) + pypp::PyList({3, 4});
    pypp::print(c);
    c += pypp::PyList({5, 6});
    pypp::print(c);
    pypp::PyList<int> d = c * 3;
    pypp::print(d);
    c *= 3;
    pypp::print(c);
    pypp::print(c[pypp::py_slice(1, 4, 1)]);
    pypp::print(c[pypp::py_slice(1, c.len() - 1, 2)]);
    pypp::print(c[pypp::py_slice(1, std::nullopt, 2)]);
    pypp::print(c[pypp::py_slice(1, std::nullopt, 1)]);
    pypp::print(c[pypp::py_slice(0, 4, 1)]);
    pypp::print(c[pypp::py_slice(0, std::nullopt, 2)]);
    pypp::print(c[pypp::py_slice(0, std::nullopt, 1)]);
    pypp::print(c[pypp::py_slice(c.len() - 2, 5, 1)]);
    pypp::print(c[pypp::py_slice(c.len() - 4, c.len() - 1, 1)]);
    pypp::print(c[pypp::py_slice(5, 1, -1)]);
    pypp::PyList<pypp::PyStr> e({pypp::PyStr("a"), pypp::PyStr("b")});
    pypp::print(e);
    pypp::PyList<int> f({});
    pypp::print(f);
    e.reserve(10);
    pypp::PyList<pypp::PyList<int>> g(
        {pypp::PyList({1, 2}), pypp::PyList({3, 4})});
    pypp::PyList<int> &g_0 = g[0];
    g_0.append(99);
    pypp::print(g);
    g[1].append(98);
    pypp::print(g);
    _test_fn(pypp::PyList({pypp::PyStr("1"), pypp::PyStr("2")}));
    e.extend(pypp::PyList({pypp::PyStr("c"), pypp::PyStr("d")}));
    pypp::print(e);
    pypp::PyList<pypp::PyStr> h({pypp::PyStr("e"), pypp::PyStr("f")});
    e.extend(std::move(h));
    pypp::print(e);
    pypp::PyList<int> i({5, 3, 1, 4, 2});
    i.sort();
    pypp::print(i);
    pypp::PyList<int> j = i[pypp::py_slice(0, std::nullopt, 1)];
    j[1] = 99;
    pypp::print(i);
    pypp::print(j);
}

} // namespace me

File: pypp_cli/test_dir/python/src/lists/lists.py
from pypp_python import mov, Valu, Ref, list_reserve


def _test_fn(arg1: Valu(list[str])):
    print(arg1)


def list_fn():
    print("LIST RESULTS:")
    # Declaration
    a: list[int] = [1, 2, 3, 4, 5]
    print(a)
    # Assigning element
    a[0] = 10
    print(a)
    # Append
    a.append(11)
    print(a)
    # Accessing - indices
    # TODO: does not work anymore. Add a built-in function `lg` that works for this.
    # print(str(a[-1]))
    # Showing reassigning a reference behaviour
    # BIG NOTE: you shouldn't do this in pypp. This results in inconsistent behavior
    #  between the Python and C++ executions. In python when you reassign the original
    #  stays the same without modification, and in C++ when you reassign the original
    #  changes as well.
    b: int = a[1]
    print(str(b))
    b = 20
    print(str(b))
    print(a)
    # Len
    print(str(len(a)))
    # reverse
    a.reverse()
    print(a)
    # count
    print(str(a.count(2)))
    # index
    print(str(a.index(2)))
    # remove
    a.remove(2)
    print(a)
    # insert
    a.insert(4, 2)
    print(a)
    # insert with mov
    ins_val: int = 2
    a.insert(4, mov(ins_val))
    print(a)
    # pop
    popped_val: int = a.pop()
    print(str(popped_val))
    print(a)
    popped_val2: int = a.pop(1)
    print(str(popped_val2))
    print(a)
    # clear
    a.clear()
    print(a)
    # lexicographical comparisons
    print(str([1, 2] == [1, 2]))
    print(str([1, 2] < [1, 2]))
    print(str([1, 2] <= [1, 2]))
    print(str([1, 2] > [1, 2]))
    print(str([1, 2] >= [1, 2]))
    print(str([1, 2] != [1, 2]))
    # concatenation
    c: list[int] = [1, 2] + [3, 4]
    print(c)
    c += [5, 6]
    print(c)
    # repetition
    d: list[int] = c * 3
    print(d)
    c *= 3
    print(c)
    # slicing
    print(c[1:4])
    print(c[1 : len(c) - 1 : 2])
    print(c[1::2])
    print(c[1:])
    print(c[:4])
    print(c[::2])
    print(c[:])
    print(c[len(c) - 2 : 5])
    print(c[len(c) - 4 : len(c) - 1])
    print(c[5:1:-1])
    # slice assignment is not supported, and a value error is raised if attempted.
    # c[1:4] = [-2, -2, -2, -2]
    # list of strings
    e: list[str] = ["a", "b"]
    print(e)
    # initialize empty
    f: list[int] = []
    print(f)
    # reserve
    list_reserve(e, 10)
    # modifying references
    g: list[list[int]] = [[1, 2], [3, 4]]
    g_0: Ref(list[int]) = g[0]
    g_0.append(99)
    print(g)
    g[1].append(98)
    print(g)
    # test passing a list inline
    _test_fn(["1", "2"])
    # extend
    e.extend(["c", "d"])
    print(e)
    h: list[str] = ["e", "f"]
    e.extend(mov(h))
    print(e)
    # sort
    i: list[int] = [5, 3, 1, 4, 2]
    i.sort()
    print(i)
    # copy
    j: list[int] = [1, 2]
    k: list[int] = j.copy()
    j.append(3)
    print(f"original: {j}, copied list: {k}")
    # copy option 2
    l: list[int] = k[:]
    k.append(4)
    print(f"original: {k}, copied list: {l}")
